piecrust-uplink/src/abi/state.rs

```diff
@@ -292,6 +296,51 @@ pub fn owner<const N: usize>(contract: ContractId) -> Option<[u8; N]> {
     }
 }
 
+/// Returns given contract's free limit, if the contract exists and if it
+/// has a free limit set.
+pub fn free_limit(contract: ContractId) -> Option<u64> {
+    let contract_id_ptr = contract.as_bytes().as_ptr();
+    unsafe {
+        match ext::free_limit(contract_id_ptr) as usize {
+            0 => None,
+            _ => with_arg_buf(|buf, _| {
+                if buf[0] == 0 {
+                    None
+                } else {
+                    let mut value_bytes = [0; LEN_U64];
+                    value_bytes.copy_from_slice(&buf[1..LEN_U64 + 1]);
+                    Some(u64::from_le_bytes(value_bytes))
+                }
+            }),
+        }
+    }
+}
+
+/// Returns given contract's free gas price hint, if the contract exists and
+/// if it has a free price hint set.
+pub fn free_price_hint(contract: ContractId) -> Option<(u64, u64)> {
+    let contract_id_ptr = contract.as_bytes().as_ptr();
+
+    unsafe {
+        match ext::free_price_hint(contract_id_ptr) as usize {
+            0 => None,
+            _ => with_arg_buf(|buf, _| {
+                if buf[0] == 0 {
+                    None
+                } else {
+                    let mut value_bytes = [0; LEN_U64];
+                    value_bytes.copy_from_slice(&buf[1..LEN_U64 + 1]);
+                    let num = u64::from_le_bytes(value_bytes);
+                    value_bytes
+                        .copy_from_slice(&buf[LEN_U64 + 1..LEN_U64 * 2 + 1]);
+                    let denom = u64::from_le_bytes(value_bytes);
+                    Some((num, denom))
+                }
+            }),
+        }
+    }
+}
```

Comment:
Exact same thing here:

```suggestion
/// Returns given contract's free gas price hint, if the contract exists and
/// if it has a free price hint set.
pub fn free_price_hint(contract: ContractId) -> Option<(u64, u64)> {
    let contract_id_ptr = contract.as_bytes().as_ptr();

    unsafe {
        match ext::free_price_hint(contract_id_ptr) {
            0 => None,
            _ => with_arg_buf(|buf, _| {
                let mut value_bytes = [0; LEN_U64];
                value_bytes.copy_from_slice(&buf[0..LEN_U64]);
                let num = u64::from_le_bytes(value_bytes);
                value_bytes
                    .copy_from_slice(&buf[LEN_U64..LEN_U64 * 2]);
                let denom = u64::from_le_bytes(value_bytes);
                Some((num, denom))
            }),
        }
    }
}
```

Reply:
same here, we have to pass option if the contract exists, so this is a double option